cli/deployment/api_client.go

```diff
@@ -11,73 +11,233 @@ import (
-		return errBuff.String(), err
+	if err != nil {
+		log.Printf("failed to start cmd: %v, output: %v", err, string(output))
+		return string(output), err
 	}
 
-	return outBuff.String(), err
+	return string(output), err
 }
 
 // GetOutputs retrive existing Deployment outputs using gcloud and store result in map[string]string
 // where "resourceName.propertyName" is key, and value is string representation of the output value.
-func GetOutputs(name string, project string) (map[string]string, error) {
-	data, err := runGCloud("deployment-manager", "manifests", "describe", "--deployment", name, "--project", project)
+func GetOutputs(project string, name string) (map[string]interface{}, error) {
+	data, err := runGCloud("deployment-manager", "manifests", "describe", "--deployment", name, "--project", project, "--format", "yaml")
 	if err != nil {
 		log.Printf("Failed to get deployment manifest: %v", err)
 		return nil, err
 	}
 	return parseOutputs(data)
 }
 
-// Create creates a new Deployment based on a Deployment object passed into it.
-func Create(deployment *Deployment) (*Deployment, error) {
+/*
+returns project id taken from local gcloud configuration
+*/
+func GCloudDefaultProjectID() (string, error) {
+	data, err := runGCloud("config", "list", "--format", "yaml")
+	if err != nil {
+		return "", err
+	}
+	out := struct {
+		Core struct {
+			Project string
+		}
+	}{}
+	err = yaml.Unmarshal([]byte(data), &out)
+	if err != nil {
+		return "", err
+	}
+	return out.Core.Project, nil
+}
+
+// Create deployment based on passed Deployment object passed into it.
+// Initialize Deployment with Outputs map in case of successful creation and error otherwise.
+func CreateOrUpdate(action string, deployment *Deployment, preview bool) (string, error) {
+	if action != ActionCreate && action != ActionUpdate {
+		log.Fatalf("action %s not in [%s,%s] for deployment: %v", action, ActionCreate, ActionUpdate, deployment)
+	}
+
 	args := []string{
 		"deployment-manager",
 		"deployments",
-		"create",
+		action,
 		deployment.config.Name,
 		"--config",
 		deployment.configFile,
 		"--project",
-		deployment.config.Project,
+		deployment.config.GetProject(),
 	}
-	_, err := runGCloud(args...)
+
+	if preview {
+		args = append(args, "--preview")
+	}
+
+	output, err := runGCloud(args...)
 	if err != nil {
-		log.Printf("Failed to create deployment: %v, error: %v", deployment, err)
-		return nil, err
+		log.Printf("failed to %s deployment: %v, error: %v", action, deployment, err)
+		return output, err
+	}
+
+	if !preview {
+		outputs, err := GetOutputs(deployment.config.GetProject(), deployment.config.Name)
+		if err != nil {
+			log.Printf("on %s action, failed to get outputs for deployment: %v, error: %v", action, deployment, err)
+			return output, err
+		}
+		deployment.Outputs = outputs
+	}
+	return output, nil
+}
+
+/*
+cancel update/create/delete preview with gcloud deployments cancel-preview command,
```

Comment:
Please use line comments as suggested previously and follow doc comment convention as described at https://golang.org/doc/effective_go.html#commentary